---
title: "Genetic background lipids"
author: 
- "DeniseSl22"
date: "01/08/2022"
output: html_notebook
---

In order to establish the potential genetic relevance of a certain modeling species in lipid metabolism, we take these steps:
1. First, we obtain human genes from three databases (Gene Ontology, Reactome, and WikiPathways) together to establish a list of relevant proteins in lipid metabolism.
2. Second, we compare these three databases together, while grouping on EC-nomenclature/LIPID MAPS ontology.
3. Third, we use the overlapping genes to compare against other modeling species (mouse, zebrafish, yeast, worm, etc.), to represent which genes can be found in which species, and might make a suitable modeling organism.

## 0. Setup
```{r}
#Package manager:
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager",repos = "http://cran.us.r-project.org")
#Package for path variable
if(!"rstudioapi" %in% installed.packages()) BiocManager::install("rstudioapi")
#Libraries required for markdown documents:
if(!"markdown" %in% installed.packages()){install.packages("markdown")}
if(!"rmarkdown" %in% installed.packages()){install.packages("rmarkdown")}

#load libraries
library(rstudioapi)

# set working environment to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
```


## 1. Relevant proteins from Databases: 

From the Quick-GO website (https://www.ebi.ac.uk/QuickGO/), we downloaded data based on the following criteria:
I. Taxon: 9606 (Homo sapiens, human).
II. Gene product type: Proteins, Reviewed (Swiss-Prot), all settings for Proteomes.
III. GO terms: lipid metabolic process (GO:0006629).

Data was downloaded on: 1st of August, with 5199 annotations, as a TSV file.
```{r}
# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

##GO terms
go_terms <- read.delim(file = "Data/QuickGO-annotations-1659356807403-20220801.tsv", header = TRUE, sep = '\t')
go_terms_unique <- unique(go_terms$SYMBOL)

# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()

```

```{r}
##Reactome
if(!"SPARQL" %in% installed.packages()){
  install.packages("SPARQL")
}
library(SPARQL)
##Connect to Endpoint WikiPathways
endpointwp <- "https://sparql.wikipathways.org/sparql"

## 1. Query metadata:
queryMetadata <-
"SELECT DISTINCT ?dataset (str(?titleLit) as ?title) ?date ?license 
WHERE {
   ?dataset a void:Dataset ;
   dcterms:title ?titleLit ;
   dcterms:license ?license ;
   pav:createdOn ?date .
 }"
 #below code should be performed first to handle the ssl certificate error
options(RCurlOptions = list(cainfo = paste0( tempdir() , "/cacert.pem" ), ssl.verifypeer = FALSE))
resultsMetadata <- SPARQL(endpointwp,queryMetadata,curl_args=list(useragent=R.version.string))
showresultsMetadata <- resultsMetadata$results
remove(queryMetadata, resultsMetadata)

## 2. Query Reactome lipid PWs, through HGNC symbols:
queryReactome <-
"PREFIX cur:            <http://vocabularies.wikipathways.org/wp#Curation:> #to differentiate between Reactome and WP PWs.

select distinct (str(?wpid) as ?pathway) (count(distinct ?lipidID) AS ?LipidsInPWs) (fn:substring(?hgncId,37) as ?HGNC) 
where {
  ?pathwayRes a wp:Pathway ;                   #Define what is a pathway
              wp:organismName 'Homo sapiens' ; #Filter pathways on species Human
              dcterms:identifier ?wpid ;       #Obtain identifier of pathway
              wp:ontologyTag cur:Reactome_Approved .  #Reactome: Reactome_Approved . WP: Curation:AnalysisCollection
  
  ?metabolite wp:bdbLipidMaps ?lipidID ;       #Find the LIPID MAPS identifier for a certain metabolite
              dcterms:isPartOf ?pathwayRes .   #Connect lipid to PW 
    
  
  ?protein wp:bdbHgncSymbol ?hgncId ;          #Find proteins based on HGNC symbol
           dcterms:isPartOf ?pathwayRes .      #Connect protein to PW
}
ORDER BY DESC(?LipidsInPWs)
"
##Execute query:
resultsReactome <- SPARQL(endpointwp,queryReactome,curl_args=list(useragent=R.version.string))
showresultsReactome <- resultsReactome$results
remove(queryReactome, resultsReactome)

##Only retain PWs with 4 or more lipids involved, and create string of unieuq HGNC symbols.
reactome_terms <- showresultsReactome[showresultsReactome$LipidsInPWs>=4, ]
reactome_terms_unique <- unique(reactome_terms$HGNC)
```

```{r}
##WikiPathways
## 3. Query WikiPathways lipid PWs, through HGNC symbols:
queryWikiPathways <-
"PREFIX cur:            <http://vocabularies.wikipathways.org/wp#Curation:> #to differentiate between Reactome and WP PWs.

select distinct (str(?wpid) as ?pathway) (count(distinct ?lipidID) AS ?LipidsInPWs) (fn:substring(?hgncId,37) as ?HGNC) 
where {
  ?pathwayRes a wp:Pathway ;                   #Define what is a pathway
              wp:organismName 'Homo sapiens' ; #Filter pathways on species Human
              dcterms:identifier ?wpid ;       #Obtain identifier of pathway
              wp:ontologyTag cur:AnalysisCollection .  #Reactome: Reactome_Approved . WP: Curation:AnalysisCollection
  
  ?metabolite wp:bdbLipidMaps ?lipidID ;       #Find the LIPID MAPS identifier for a certain metabolite
              dcterms:isPartOf ?pathwayRes .   #Connect lipid to PW 
    
  
  ?protein wp:bdbHgncSymbol ?hgncId ;          #Find proteins based on HGNC symbol
           dcterms:isPartOf ?pathwayRes .      #Connect protein to PW
}
ORDER BY DESC(?LipidsInPWs)
"
##Execute query:
resultsWikiPathways <- SPARQL(endpointwp,queryWikiPathways,curl_args=list(useragent=R.version.string))
showresultsWikiPathways <- resultsWikiPathways$results
remove(queryWikiPathways, resultsWikiPathways)

##Only retain PWs with 4 or more lipids involved, and create string of unieuq HGNC symbols.
wikipathways_terms <- showresultsWikiPathways[showresultsWikiPathways$LipidsInPWs>=4, ]
wikipathways_terms_unique <- unique(wikipathways_terms$HGNC)

```

## 2. Compare databases

From the HGNC biomart webiste (http://biomart.genenames.org/), we downloaded data based on the following criteria:
I. Start search by selecting "gene"
II. Status: Approved
III. Attributes: HGNC ID, Status, Approved Symbol, Approved Name.
IV. Protein resources: Enzyme (EC) ID.

HGNC to EC-code data was downloaded on: 1st of August, with 43295 lines, as a TSV file.
```{r}
# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

# Read the HGNC to EC-code file:
EC_terms <- read.delim(file = "Data/HGNC_biomart.txt", header = TRUE, sep = '\t', na.strings = "")

# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()

# Add 'unclassified' iso 'NA' values
EC_terms <- replace(EC_terms, is.na(EC_terms), "unclassified")
# Keep only relevant columns:
EC_terms_HGNC <- EC_terms[, c("Approved.symbol","Enzyme..EC..ID")]

```

Link the EC codes to HGNC symbols for each database, and combine the data in one dataframe:
```{r}
## Find HGNC symbols that are not recognized, for data curation:
EC_terms_GO_not <-  setdiff(go_terms_unique, EC_terms_HGNC$Approved.symbol) 
EC_terms_Reactome_not <-  setdiff(reactome_terms_unique, EC_terms_HGNC$Approved.symbol) 
EC_terms_WikiPathways_not <-  setdiff(wikipathways_terms_unique, EC_terms_HGNC$Approved.symbol) 

##TODO: Map these HGNC IDs to automatically updated to new entries through BridgeDb using secondary to primary mapping.

# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

## Find these HGNC symbols in the 'multi-symbol-checker' (https://www.genenames.org/tools/multi-symbol-checker/), 
## Manually remove the first row 'sep=', otherwise the file cannot be read in properly
## and read in CSV download:
HGNC_mapping_terms <- read.csv(file = "Data/hgnc-symbol-check.csv", header = TRUE, na.strings = "")

# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()

#Create new array to compare changes later:
go_terms_new <- go_terms_unique
reactome_terms_new <- reactome_terms_unique
wikipathways_terms_new <- wikipathways_terms_unique

##Replace the symbols if 'match.type' == 	'Previous symbol', otherwise remove the entry
#GO
for (i in 1:nrow(HGNC_mapping_terms)) {
    if (HGNC_mapping_terms$Match.type[i] == 'Previous symbol') {
    #Update entry
    go_terms_new <- replace(go_terms_new, go_terms_new == HGNC_mapping_terms$Input[i], HGNC_mapping_terms$Approved.symbol[i])
    }else if(HGNC_mapping_terms$Match.type[i] != 'Previous symbol'){
    #Remove entry
    go_terms_new <-  replace(go_terms_new, go_terms_new == HGNC_mapping_terms$Input[i], 'remove')
    go_terms_new <- go_terms_new[!go_terms_new %in% 'remove']
    }
}

#Reactome
for (i in 1:nrow(HGNC_mapping_terms)) {
    if (HGNC_mapping_terms$Match.type[i] == 'Previous symbol') {
    #Update entry
    reactome_terms_new <- replace(reactome_terms_new, reactome_terms_new == HGNC_mapping_terms$Input[i], HGNC_mapping_terms$Approved.symbol[i])
    }else if(HGNC_mapping_terms$Match.type[i] != 'Previous symbol'){
    #Remove entry
    reactome_terms_new <- replace(reactome_terms_new, reactome_terms_new == HGNC_mapping_terms$Input[i], 'remove')
    reactome_terms_new <- reactome_terms_new[!reactome_terms_new %in% 'remove']
    }
}

#WikiPathways
for (i in 1:nrow(HGNC_mapping_terms)) {
    if (HGNC_mapping_terms$Match.type[i] == 'Previous symbol') {
    #Update entry
    wikipathways_terms_new <- replace(wikipathways_terms_new, wikipathways_terms_new == HGNC_mapping_terms$Input[i], HGNC_mapping_terms$Approved.symbol[i])
    }else if(HGNC_mapping_terms$Match.type[i] != 'Previous symbol'){
    #Remove entry
    wikipathways_terms_new <- replace(wikipathways_terms_new, wikipathways_terms_new == HGNC_mapping_terms$Input[i], 'remove')
    wikipathways_terms_new <- wikipathways_terms_new[!wikipathways_terms_new %in% 'remove']
    }
}

```

Visualize overlap in Venn diagram (to play around with the settings, before saving the Figure as a file):
```{r}
if(!"VennDiagram" %in% installed.packages()){install.packages("VennDiagram")}
if(!"RColorBrewer" %in% installed.packages()){install.packages("RColorBrewer")}
# Load library
library(VennDiagram)
# Prepare a palette of 3 colors with R colorbrewer:
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")
##Ignore log messages:
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

# Chart
figure1 <- venn.diagram(
        x = list(go_terms_new, reactome_terms_new, wikipathways_terms_new),
        category.names = c("Gene Ontology" , "Reactome" , "WikiPathways"),
        filename = NULL, 
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 450 , 
        width = 300 , 
        resolution = 500,
        compression = "lzw",

        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)
grid.newpage()
grid.draw(figure1)

```
Save Venn diagram as a Figure:
```{r}
# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

# Chart
venn.diagram(
        x = list(go_terms_new, reactome_terms_new, wikipathways_terms_new),
        category.names = c(" Gene Ontology" , "Reactome" , "WikiPathways"),
        filename = 'Figures/Overlap_GO_Re_WP.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 250 , 
        width = 250 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .4,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.22,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)
# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()
```

Connect data to EC codes:
```{r}
if(!"dplyr" %in% installed.packages()){
  install.packages("dplyr")
}
library(dplyr)

##Link data from three databases to EC-codes:
EC_terms_GO <-  EC_terms_HGNC[EC_terms_HGNC$Approved.symbol %in% go_terms_new, ]
EC_terms_Reactome <-  EC_terms_HGNC[EC_terms_HGNC$Approved.symbol %in% reactome_terms_new, ]
EC_terms_WikiPathways <-  EC_terms_HGNC[EC_terms_HGNC$Approved.symbol %in% wikipathways_terms_new, ]

##Combine data
combined_data12 <- merge(x = EC_terms_GO, y = EC_terms_Reactome, by = "Approved.symbol", all.x = TRUE, all.y=TRUE, sort = TRUE)
combined_data_all <- merge(x = combined_data12 , y = EC_terms_WikiPathways, by = "Approved.symbol", all.x = TRUE, all.y=TRUE, sort = TRUE)

#Rename the columns:
colnames(combined_data_all) <- c("HGNC", "GO", "Reactome", "WikiPathways")
```

Convert data to adjacency matrix:
```{r}
#create data backup:
combined_data_noDuplicates <- combined_data_all
# Remove duplicate HGNC terms, since several HGNC terms have multiple EC-codes (e.g. AADAT has EC-code 2.6.1.7 and 2.6.1.39):
combined_data_noDuplicates <- combined_data_noDuplicates %>% distinct(combined_data_noDuplicates$HGNC , .keep_all = TRUE)
#Remove last column added by distinct method
combined_data_noDuplicates <- subset(combined_data_noDuplicates, select = -c(5) )
##Count entries for each column, to see if this matches with new terms and we're not forgetting some proteins along the way!
length(go_terms_new) == sum(!is.na(combined_data_noDuplicates$GO))
length(reactome_terms_new) == sum(!is.na(combined_data_noDuplicates$Reactome))
length(wikipathways_terms_new) == sum(!is.na(combined_data_noDuplicates$WikiPathways))

# Sort the columns based on HGNC symbol:
combined_data_noDuplicates <- combined_data_noDuplicates[order(combined_data_noDuplicates$HGNC),]

##Unify the EC entries to their class name:
# GO
combined_data_noDuplicates <- combined_data_noDuplicates%>%mutate(GO_class = case_when(
  startsWith(combined_data_noDuplicates$GO, '1.') ~ 'G1',
  startsWith(combined_data_noDuplicates$GO, '2.') ~ 'G2',
  startsWith(combined_data_noDuplicates$GO, '3.') ~ 'G3',
  startsWith(combined_data_noDuplicates$GO, '4.') ~ 'G4',
  startsWith(combined_data_noDuplicates$GO, '5.') ~ 'G5',
  startsWith(combined_data_noDuplicates$GO, '6.') ~ 'G6',
  startsWith(combined_data_noDuplicates$GO, '7.') ~ 'G7',
  is.na(combined_data_noDuplicates$GO) ~ combined_data_noDuplicates$GO, 
  TRUE ~ 'GO_U'
))

# Reactome (Re)
combined_data_noDuplicates <- combined_data_noDuplicates%>%mutate(Re_class = case_when(
  startsWith(combined_data_noDuplicates$Reactome, '1.') ~ 'R1',
  startsWith(combined_data_noDuplicates$Reactome, '2.') ~ 'R2',
  startsWith(combined_data_noDuplicates$Reactome, '3.') ~ 'R3',
  startsWith(combined_data_noDuplicates$Reactome, '4.') ~ 'R4',
  startsWith(combined_data_noDuplicates$Reactome, '5.') ~ 'R5',
  startsWith(combined_data_noDuplicates$Reactome, '6.') ~ 'R6',
  startsWith(combined_data_noDuplicates$Reactome, '7.') ~ 'R7',
  is.na(combined_data_noDuplicates$Reactome) ~ combined_data_noDuplicates$Reactome, 
  TRUE ~ 'Re_U'
))

# WikiPathways (WP)
combined_data_noDuplicates <- combined_data_noDuplicates%>%mutate(WP_class = case_when(
  startsWith(combined_data_noDuplicates$WikiPathways, '1.') ~ 'W1',
  startsWith(combined_data_noDuplicates$WikiPathways, '2.') ~ 'W2',
  startsWith(combined_data_noDuplicates$WikiPathways, '3.') ~ 'W3',
  startsWith(combined_data_noDuplicates$WikiPathways, '4.') ~ 'W4',
  startsWith(combined_data_noDuplicates$WikiPathways, '5.') ~ 'W5',
  startsWith(combined_data_noDuplicates$WikiPathways, '6.') ~ 'W6',
  startsWith(combined_data_noDuplicates$WikiPathways, '7.') ~ 'W7',
  is.na(combined_data_noDuplicates$WikiPathways) ~ combined_data_noDuplicates$WikiPathways, 
  TRUE ~ 'WP_U'
))

if(!"reshape2" %in% installed.packages()){
  install.packages("reshape2")
}
if(!"reshape" %in% installed.packages()){
  install.packages("reshape")
}
library(reshape2) 
library(reshape) 

#Keep only relevant columns
GO_Re <- combined_data_noDuplicates[,c(5,6)]
GO_WP <- combined_data_noDuplicates[,c(5,7)]
WP_Re <- combined_data_noDuplicates[,c(6,7)]

#Melt data to obtain an edge list
melt_data_GO_Re <- na.omit(melt(GO_Re, id = c('GO_class'))) 
melt_data_GO_WP <- na.omit(melt(GO_WP, id = c('GO_class'))) 
melt_data_WP_Re <- na.omit(melt(WP_Re, id = c('Re_class'))) 

#Merge data back together
lst <- mget(ls(pattern='^melt_data_'))
list2env(lapply(lst,`[`,-2), envir=.GlobalEnv)
names(melt_data_WP_Re)[1] <- names(melt_data_GO_Re)[1]
merged_comparison <-  do.call("rbind", list(melt_data_GO_Re, melt_data_GO_WP, melt_data_WP_Re))
names(merged_comparison) <- c('Database1','DataBase2')

#Filter out the Unclassified interactions, since these clutter the visualization:
merged_comparison <- dplyr::filter(merged_comparison, !grepl('U', Database1))

##Calculate occurrence (edge weight)
#if(!"plyr" %in% installed.packages()){
#  install.packages("plyr")
#}
#library(plyr)
#merged_comparison_weight <- ddply(merged_comparison, .(Database1,DataBase2), nrow)
#names(merged_comparison_weight)[3] <- 'weight'
##Count unique entries in dataset before constructing the matrix size:
#unique_edges <-  merged_comparison_weight %>% 
#  select(Database1, DataBase2) %>% 
#  t %>% c %>% unique

##Use the igraph library to calculate the adjacency matrix:
if(!"igraph" %in% installed.packages()){install.packages("igraph")}
suppressPackageStartupMessages(library(igraph))
library(igraph)
adjacencyData <- as.matrix(get.adjacency(graph.data.frame(merged_comparison)))

#Order the labels of the data:
adjacencyData_ordered_row <- adjacencyData[order(rownames(adjacencyData)),]
adjacencyData_ordered <- adjacencyData_ordered_row[,order(colnames(adjacencyData_ordered_row))]

##Create a chord diagram to visualize overlap:
# Charge the circlize library
if(!"circlize" %in% installed.packages()){install.packages("circlize")}
suppressPackageStartupMessages(library(circlize))
library(circlize)
 
# Make the circular plot
chordDiagram(adjacencyData_ordered, transparency = 0.5)
circos.clear()

```

Update the visualization by grouping the data per database, and using the same colors from the Venn Diagram (to play around with the settings, before saving the Figure as a file):
```{r}
#Group the labels for WP, Reactome, GO; for better readability
nm = unique(unlist(dimnames(adjacencyData_ordered))) #Take out all labels as individual entries
group = structure(gsub("_\\d", "", nm), names = nm) #Check each entry against regex, remove '_(number)', to keep database name only.

#Apply a color scheme for matching groups
grid.col = structure(rep(c("#ffa600", "#ff6361", "#bc5090", "#58508d", "#003f5c", "#000000"), 3))

#Visualize groups better:
chordDiagram(adjacencyData_ordered, group = group, grid.col = grid.col, transparency = 0.5)
circos.clear()

##Add multi-group labels:

#The color grid can also be defined manually, to show each EC class in the same color. 
#Use circos.info() to get the order; the color can also be add without specifying the names, however, If it is has no name index, the order of grid.col is assumed to have the same order as sectors
#sectors: "G1" "G2" "G3" "G4" "G5" "G6" "R1" "R2" "R3" "R4" "R5" "R6" "W1" "W2" "W3" "W4" "W5" "W6"
#grid.col = c(G1 = "#ffa600", G2 = "#ff6361", G3 = "#bc5090", G4 = "#58508d", G5 = "#003f5c", G6 = "#000000",
#    R1 = "#ffa600", R2 = "#ff6361", R3 = "#bc5090", R4 = "#58508d", R5 = "#003f5c", R6 = "#000000",
#    W1 = "#ffa600", W2 = "#ff6361", W3 = "#bc5090", W4 = "#58508d", W5 = "#003f5c", W6 = "#000000")

grid.col = structure(rep(c("#ffa600", "#ff6361", "#bc5090", "#58508d", "#003f5c", "#000000"), 3))

group = structure(gsub("_\\d", "", nm), names = nm)
chordDiagram(adjacencyData_ordered, group = group, grid.col = grid.col, transparency = 0.5,
    annotationTrack = c("grid", "axis"),
     annotationTrackHeight = c(0.1, 0.1),
    preAllocateTracks = list(
        track.height = mm_h(4),
        track.margin = c(mm_h(4), 0)
))
p <- circos.track(track.index = 1, panel.fun = function(x, y) {
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.text(mean(xlim), mean(ylim), sector.index, cex = 0.5,  facing = "clockwise", niceFacing = TRUE)
}, bg.border = NA)

highlight.sector(c(paste0("R", 1:6)), track.index = 2, col = myCol[2], 
    text = "Reactome", cex = 0.7, text.col = "white", niceFacing = TRUE)
highlight.sector(c(paste0("W", 1:6)), track.index = 2, col =  myCol[3], 
    text = "WikiPathways", cex = 0.7, text.col = "white", niceFacing = TRUE)
highlight.sector(c(paste0("G", 1:6)), track.index = 2, col = myCol[1], 
    text = "Gene Ontology", cex = 0.7, text.col = "white", niceFacing = TRUE)

#Add legend with EC code classification, reusing the colors for the diagram as defined above by the parameter "grid.col".
col = c("1: Oxidoreductase" = grid.col[1],
        "2: Transferases" = grid.col[2],
        "3: Hydrolases" = grid.col[3],
        "4: Lyases" = grid.col[4],
        "5: Isomerases" = grid.col[5],
        "6: Ligase" = grid.col[6]) # EC 7: Translocases --> Not found.

legend("bottomright", pch = 15, legend = names(col), col = col, title = "Enzyme Classification", cex = 0.8) 

circos.clear()

```

Save chord diagram as a Figure:
```{r}
# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

##Save the figure in PNG file suited for paper:
png(filename="Figures/overlap_Genes_chordDiagram.png", width = 430, height = 280, units = "px") 

grid.col = structure(rep(c("#ffa600", "#ff6361", "#bc5090", "#58508d", "#003f5c", "#000000"), 3))
group = structure(gsub("_\\d", "", nm), names = nm)
chordDiagram(adjacencyData_ordered, group = group, grid.col = grid.col, transparency = 0.5,
    annotationTrack = c("grid", "axis"),
     annotationTrackHeight = c(0.1, 0.1),
    preAllocateTracks = list(
        track.height = mm_h(4),
        track.margin = c(mm_h(4), 0)
))
circos.track(track.index = 1, panel.fun = function(x, y) {
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.text(mean(xlim), mean(ylim), sector.index, cex = 0.7,  facing = "clockwise", niceFacing = TRUE)
}, bg.border = NA)

highlight.sector(c(paste0("R", 1:6)), track.index = 2, col = myCol[2], 
    text = "Reactome", cex = 0.7, text.col = "white", niceFacing = TRUE)
highlight.sector(c(paste0("W", 1:6)), track.index = 2, col =  myCol[3], 
    text = "WikiPathways", cex = 0.7, text.col = "white", niceFacing = TRUE)
highlight.sector(c(paste0("G", 1:6)), track.index = 2, col = myCol[1], 
    text = "Gene Ontology", cex = 0.7, text.col = "white", niceFacing = TRUE)

#Add legend with EC code classification, reusing the colors for the diagram as defined above by the parameter "grid.col".
col = c("1: Oxidoreductase" = grid.col[1],
        "2: Transferases" = grid.col[2],
        "3: Hydrolases" = grid.col[3],
        "4: Lyases" = grid.col[4],
        "5: Isomerases" = grid.col[5],
        "6: Ligase" = grid.col[6]) # EC 7: Translocases --> Not found.

legend("bottomright", pch = 15, legend = names(col), col = col, title = "Enzyme Classification", cex = 0.7) 

circos.clear()

dev.off() 

# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()
```
## 3. Model species comparison 

Obtain overlapping genes between two and three databases:
```{r}
#Calculate the overlap
overlap <- calculate.overlap(
  x = list(go_terms_new, reactome_terms_new, wikipathways_terms_new)
)

#The first four values in the list (in the case of comparing 3 databases) are relevant to keep.
top4_overlap <- c(overlap[[1]], overlap[[2]], overlap[[3]], overlap[[4]])

```

For these genes annotated with HGNC symbols, find homologous in different species.
(see https://shiring.github.io/genome/2016/12/11/homologous_genes_post for inspiration; note that this script is from 2016 and several function do not work directly anymore, e.g. the getLDS function requires a archived (older) version of Ensembl, see https://support.bioconductor.org/p/9143401/#9143687.)
```{r}
if(!"biomaRt" %in% installed.packages()){install.packages("biomaRt")}
suppressPackageStartupMessages(library(biomaRt))
library(biomaRt)

#Find names for available species in Ensembl:
ensembl =  biomaRt::useMart("ensembl")
datasets <- listDatasets(ensembl)

#Subset relevant for this paper:
#Saccharomyces cerevisiae, Neurosporea crassa (not part of Ensembl download), Aspergillus nidulans(not part of Ensembl download), Caenorhabditis elegans, Drosophila melanogaster, Mus musculus, Arabidopsis thaliananot part of Ensembl download).
speciesPaper <- c("hsapiens_gene_ensembl", "celegans_gene_ensembl", "dmelanogaster_gene_ensembl", "mmusculus_gene_ensembl", "scerevisiae_gene_ensembl")

human.mart <- biomaRt::useMart(host="https://dec2021.archive.ensembl.org", "ENSEMBL_MART_ENSEMBL", dataset=speciesPaper[1])
celegans.mart <- biomaRt::useMart(host="https://dec2021.archive.ensembl.org", "ENSEMBL_MART_ENSEMBL", dataset=speciesPaper[2])
dmelanogaster.mart <- biomaRt::useMart(host="https://dec2021.archive.ensembl.org", "ENSEMBL_MART_ENSEMBL", dataset=speciesPaper[3])
mouse.mart <- biomaRt::useMart(host="https://dec2021.archive.ensembl.org", "ENSEMBL_MART_ENSEMBL", dataset=speciesPaper[4])
scerevisiae.mart <- biomaRt::useMart(host="https://dec2021.archive.ensembl.org", "ENSEMBL_MART_ENSEMBL", dataset=speciesPaper[5])

human2celegansLipids.df <- biomaRt::getLDS(attributes=c("external_gene_name","ensembl_gene_id"),filters=c("external_gene_name"),values=top4_overlap,mart=human.mart,attributesL=c("external_gene_name","ensembl_gene_id"),martL=celegans.mart)

human2dmelanogasterLipids.df <- biomaRt::getLDS(attributes=c("external_gene_name","ensembl_gene_id"),filters=c("external_gene_name"),values=top4_overlap,mart=human.mart,attributesL=c("external_gene_name","ensembl_gene_id"),martL=dmelanogaster.mart)

human2mouseLipids.df <- biomaRt::getLDS(attributes=c("external_gene_name","ensembl_gene_id"),filters=c("external_gene_name"),values=top4_overlap,mart=human.mart,attributesL=c("external_gene_name","ensembl_gene_id"),martL=mouse.mart)

human2scerevisiaeLipids.df <- biomaRt::getLDS(attributes=c("external_gene_name","ensembl_gene_id"),filters=c("external_gene_name"),values=top4_overlap,mart=human.mart,attributesL=c("external_gene_name","ensembl_gene_id"),martL=scerevisiae.mart)

```

Merge IDs where more then one mapping were found in one dataframe:
```{r}
#Create a dataframe with all gene names:
lipidGenes <- data.frame(top4_overlap)
colnames(lipidGenes) <- "Gene.name"

#Count the occurence of each human gene name in the hmology data, to find 1-to-many-mappings.
mouse.occurrence <- human2mouseLipids.df %>% count(Gene.name) 
celegans.occurrence <- human2celegansLipids.df %>% count(Gene.name) 
dmelanogaster.occurrence <- human2dmelanogasterLipids.df %>% count(Gene.name) 
scerevisiae.occurrence  <- human2scerevisiaeLipids.df %>% count(Gene.name) 

#Merge the data in one dataframe:
lipidGenes <- left_join(lipidGenes, mouse.occurrence, by = "Gene.name") %>%
              left_join(., celegans.occurrence, by='Gene.name') %>%
              left_join(., dmelanogaster.occurrence, by='Gene.name') %>% 
              left_join(., scerevisiae.occurrence, by='Gene.name') %>% 
                mutate_each(funs(replace(., which(is.na(.)), 0)))
colnames(lipidGenes) <- c("Gene.name", "mouse.occurrence", "celegans.occurrence", "dmelanogaster.occurrence", "scerevisiae.occurrence")

##Clear up irrelevant data to free up memory in workspace:
#rm(list=setdiff(ls(), c("lipidGenes", "top4_overlap")))
```

Visualize the homology mapped data
```{r}
##Create summary of occurrence for whole gene list, and add column with species name
occurrence.Mouse.Summary <- data.frame(table(lipidGenes$mouse.occurrence))
occurrence.Mouse.Summary$species <- "mouse"

occurrence.celegans.Summary <- data.frame(table(lipidGenes$celegans.occurrence))
occurrence.celegans.Summary$species <- "celegans"

occurrence.dmelanogaster.Summary <- data.frame(table(lipidGenes$dmelanogaster.occurrence))
occurrence.dmelanogaster.Summary$species <- "dmelanogaster"

occurrence.scerevisiae.Summary <- data.frame(table(lipidGenes$scerevisiae.occurrence))
occurrence.scerevisiae.Summary$species <- "scerevisiae"

#Merge the data in one dataframe:
Data_overview <- bind_rows(occurrence.Mouse.Summary, occurrence.celegans.Summary, occurrence.dmelanogaster.Summary, occurrence.scerevisiae.Summary) %>% arrange(Var1)
colnames(Data_overview) <- c("Mappings", "Occurrence", "species")

#Count Missing IDs:
missingOccurrence.mouse <- filter(Data_overview, Mappings==0 & species%in% ("mouse"))
missingOccurrence.celegans <- filter(Data_overview, Mappings==0 & species%in% ("celegans"))
missingOccurrence.dmelanogaster <- filter(Data_overview, Mappings==0 & species%in% ("dmelanogaster"))
missingOccurrence.scerevisiae <- filter(Data_overview, Mappings==0 & species%in% ("scerevisiae"))

##Summarize Missing mappings (NA) for these species:
paste0("No mappings were found for ", missingOccurrence.mouse[2], " M. musculus genes (", round((missingOccurrence.mouse[2]/length(top4_overlap)*100), digits = 1), "% of the total gene list), ",
       missingOccurrence.celegans[2], " C. elegans genes (" , round((missingOccurrence.celegans[2]/length(top4_overlap)*100), digits = 1), "%), ",
       missingOccurrence.dmelanogaster[2], " D. melanogaster genes (" , round((missingOccurrence.dmelanogaster[2]/length(top4_overlap)*100), digits = 1), "%), and ",
       missingOccurrence.scerevisiae[2], " S. cerevisiae genes (" , round((missingOccurrence.scerevisiae[2]/length(top4_overlap)*100), digits = 1), "%)."
       )

#Count 1-to-1 mappings:
one_Occurrence.mouse <- filter(Data_overview, Mappings==1 & species%in% ("mouse"))
one_Occurrence.celegans <- filter(Data_overview, Mappings==1 & species%in% ("celegans"))
one_Occurrence.dmelanogaster <- filter(Data_overview, Mappings==1 & species%in% ("dmelanogaster"))
one_Occurrence.scerevisiae <- filter(Data_overview, Mappings==1 & species%in% ("scerevisiae"))

##Summarize 1-to-1 mappings for these species:
paste0("One-to-one mappings were found for ", one_Occurrence.mouse[2], " M. musculus genes (", round((one_Occurrence.mouse[2]/length(top4_overlap)*100), digits = 1), "% of the total gene list), ",
       one_Occurrence.celegans[2], " C. elegans genes (" , round((one_Occurrence.celegans[2]/length(top4_overlap)*100), digits = 1), "%), ",
       one_Occurrence.dmelanogaster[2], " D. melanogaster genes (" , round((one_Occurrence.dmelanogaster[2]/length(top4_overlap)*100), digits = 1), "%), and ",
       one_Occurrence.scerevisiae[2], " S. cerevisiae genes (" , round((one_Occurrence.scerevisiae[2]/length(top4_overlap)*100), digits = 1), "%)."
       )

##Remove occurrence 1 for better visualization of other data:
#Data_overview$Mappings <- as.numeric(Data_overview$Mappings)
More_mappings <- filter(Data_overview, Mappings != 1 & Mappings != 0 )
More_mappings$Mappings <- as.numeric(as.character(More_mappings$Mappings))
More_mappings <- More_mappings[order(More_mappings$Mappings),]
##Below should lead to the same results as above two lines, but is showing to be inconsitent...
#More_mappings <- More_mappings[order(as.numeric(as.character(More_mappings$Mappings))),]

##Count total 1-to-many mappings:
many_Occurrence.mouse <- filter(More_mappings, species%in% ("mouse"))
many_Occurrence.celegans <- filter(More_mappings, species%in% ("celegans"))
many_Occurrence.dmelanogaster <- filter(More_mappings, species%in% ("dmelanogaster"))
many_Occurrence.scerevisiae <- filter(More_mappings, species%in% ("scerevisiae"))

##Summarize 1-to-many mappings for these species:
paste0("One-to-many mappings were found for ", sum(many_Occurrence.mouse$Occurrence), " M. musculus genes (", round((sum(many_Occurrence.mouse$Occurrence)/length(top4_overlap)*100), digits = 1), "% of the total gene list), ",
      sum(many_Occurrence.celegans$Occurrence), " C. elegans genes (" , round((sum(many_Occurrence.celegans$Occurrence)/length(top4_overlap)*100), digits = 1), "%) ",
      sum(many_Occurrence.dmelanogaster$Occurrence), " D. melanogaster genes (" , round((sum(many_Occurrence.dmelanogaster$Occurrence)/length(top4_overlap)*100), digits = 1), "%), and ",
      sum(many_Occurrence.scerevisiae$Occurrence), " S. cerevisiae genes (" , round((sum(many_Occurrence.scerevisiae$Occurrence)/length(top4_overlap)*100), digits = 1), "%)."
       )

#Download and install the relevant packages:
if(!"ggplot2" %in% installed.packages()){install.packages("ggplot2")}
suppressPackageStartupMessages(library(ggplot2))
library(ggplot2)
if(!"hrbrthemes" %in% installed.packages()){install.packages("hrbrthemes")}
suppressPackageStartupMessages(library(hrbrthemes))
library(hrbrthemes)
if(!"ggbreak" %in% installed.packages()){install.packages("ggbreak")}
suppressPackageStartupMessages(library(ggbreak))
library(ggbreak)

# A scatterplot with color depending on Species
ggplot(More_mappings, aes(x = Mappings, y = Occurrence, color = species, shape=species, size=species)) + 
    geom_point(size=3) +
    theme_ipsum() + #Other themes: theme_ipsum: Arial Narrow; theme_ipsum_es: Econ Sans Condensed; theme_ipsum_rc: Roboto Condensed; theme_ipsum_ps: IBM Plex Sans font; theme_ipsum_pub: Public Sans; theme_ipsum_tw: Titilium Web; theme_modern_rc: Roboto Condensed dark theme; theme_ft_rc: Dark theme based on FTâ€™s dark theme (Roboto Condensed). Not available:  theme_ipsum_gs: Goldman Sans Condensed;
    theme(axis.title.x = element_text(size = 14, face = "bold", hjust = 0.5) ,
    axis.title.y = element_text(size = 14, face = "bold", hjust = 0.5) )+
    scale_x_continuous(breaks = seq(0, 190, 5), limits = c(0, 185)) +
    scale_x_break(c(55, 178)) 

```

Save the plot above as a Figure suitable for publication:
```{r}
# Obtain Working Directory to find data
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
getwd()
setwd('..')
work_DIR <- getwd()

ggsave(
  filename="Figures/Genes_Mappings_Occurrence.png",
  plot = last_plot(),
  device = NULL,
  path = NULL,
  scale = 3,
  width = 800,
  height = 500,
  units = "px",
  dpi = 300,
  limitsize = TRUE, 
  bg = "#FFFFFF",
)

# Set Working Directory back to current folder
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
work_DIR <- getwd()

```


Show how groups of enzymes (through EC-code) are covered in different species?
```{r}
## Add EC codes to dataframe
#colnames(EC_terms_HGNC) <- c("Gene.name", "EC")
##Merge EC code to dataframe
#lipidGenes <- left_join(lipidGenes, EC_terms_HGNC, by = "Gene.name")

# Update dataframe to
#lipidGenes <- lipidGenes%>%mutate(EC_class = case_when(
#  startsWith(lipidGenes$EC, '1.') ~ 'EC1',
#  startsWith(lipidGenes$EC, '2.') ~ 'EC2',
#  startsWith(lipidGenes$EC, '3.') ~ 'EC3',
#  startsWith(lipidGenes$EC, '4.') ~ 'EC4',
#  startsWith(lipidGenes$EC, '5.') ~ 'EC5',
#  startsWith(lipidGenes$EC, '6.') ~ 'EC6',
#  startsWith(lipidGenes$EC, '7.') ~ 'EC7',
#  is.na(lipidGenes$EC) ~  lipidGenes$EC, #NA,
#  TRUE ~ 'Unclassified'
#))

```

